### PR TITLE
WIP: Rearrange tests into a directory structure

### PR DESCRIPTION
**Pull request by ChrisJefferson**
This adds a new function `TestDirectory`, and rearranges the gap tests to use it. comments welcome!

### CODE REVIEW

**Comment by fingolfin** [comment]
Yay for the idea! As to the patch series: The commits are not always nicely named. Are you planing on squashing some of them, or at least rewriting them to have meaningful commit messages?

**Comment by alex-konovalov** [comment]
Please don't merge this anyone before agreeing on a date: first we have to clean up existing failures, or at least to identify them all, and then switch tests. We use test logs to extract runtime information and detect failures automatically, so some things may be very sensitive to any changes.

**Comment by ChrisJefferson** [comment]
Yes, I plan on rearranging this by file, and squashing all around. I expect to have to do a few more fixes to shake out problems as well.

**Comment by ChrisJefferson** [comment]
The runtime information (by which I assume you mean time things take to run?) Is probably going to get hopelessly broken, because so many things are a mess at the moment (some tests never run, tests run differently depending on how you run them). The failures should (hopefully!) get much easier. The plan is there should no longer be a need for any 'grepping' of the output to see what failures occurred - GAP will just return the correct value.

**Comment by alex-konovalov** [comment]
I will clone your branch and have a look - had no time to do this yet. Just a few words on the current setting: besides grepping for diffs to detect failures, I am extracting third column with runtimes for a dozen of individual tests and the whole testsuite:
```
testing: /home/hudson/hudson/workspace/GAP-major-release-test/GAPCOPTS/32build\
/GAPGMP/nogmp/GAPTARGET/standard/label/graupius/gap4r8/tst/grpconst.tst
grpconst.tst         1154547         179763
-------------------------------------------
total                  61121        2365496
```
I do not care about GAPstones (the 2nd column) and perhaps they may gone, but I'd like to learn how the runtime should be extracted and passed to Jenkings jobs to build plots. We may just start to accumulate new history of runtimes after the switch to the new setup, if they will not be consistent with the previous logs any more.

**Comment by alex-konovalov** [comment]
Possible improvement: use `tee` to redirect output to a file and show it on the screen at the same time.

This will not only help for a long running test like `make teststandard` to be more informative and give the tester a better idea at which stage the test is, but will be also be useful when running such tests in Travis, which needs some output to be generated. Thus, we will increase re-use of the code for testing.

File: tst/testtravis.g [hunk 1]
```diff
@@ -41,72 +41,9 @@ Print( "You should start GAP4 using `gap -A -x 80 -r -m 100m -o 750m'.\n",
...
-  [ "vspcmat.tst",8400000],
-  [ "vspcrow.tst",47400000],
-  [ "weakptr.tst",11500000],
-  [ "xgap.tst",1206600000],
-  [ "zlattice.tst",100000],
-  [ "zmodnz.tst",2300000],
-] ); 
+TestDirectory([ Filename(DirectoriesLibrary("tst"), "testall"),
+                Filename(DirectoriesLibrary("tst"), "testinstall")],
+                rec(exitGAP := true, stonesLimit := 180800000));
```
Comment (alex-konovalov) [inline tst/testtravis.g]: `testtravis.g` was created by @markuspf - the idea was that is is created from `testall.g` excluding from it some final tests that run a bit too long, to please Travis requiring for the tests to be run within 1 hour and don't be silent for intervals longer than 10 minutes.

Comment (ChrisJefferson) [inline tst/testtravis.g]: That's hopefully what this does -- the 'stonesLimit' only runs tests which take less than that number of stones (that limit could need reducing if some tests take too long)

**Comment by alex-konovalov** [comment]
I've added milestone GAP 4.8.0 - this would be a beta release of GAP 4.8 for package authors, and we want to provide a stable platform for testing their packages. Therefore, all disruptive changes which may influence packages should be ready prior to releasing GAP 4.8.0, or otherwise they would have to wait until GAP 4.9.

**Pull request closed without merging**